src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -447,7 +447,30 @@ internal void CompleteWriter(Exception? exception)
 
         internal void AdvanceReader(in SequencePosition consumed)
         {
-            AdvanceReader(consumed, consumed);
+            // If the reader is completed
+            if (_readerCompletion.IsCompleted)
+            {
+                ThrowHelper.ThrowInvalidOperationException_NoReadingAllowed();
+            }
+
+            long examinedIndex = consumed.GetInteger();
+            BufferSegment? examinedSegment = (BufferSegment?)consumed.GetObject();
+            if (examinedSegment != null &&
+                // Avoid the lock if we're examining the entire segment, don't need to look at the last examined index in that case
+                examinedSegment.Length - examinedIndex > 0)
+            {
+                lock (SyncObj)
```

Comment:
I think it would technically be ok to access `_lastExaminedIndex` outside of the lock as it should only be modified by read calls or the Pipe completing, both of which should be synchronized with the advance from the current read.

Reply:
I agree that this lock isn't necessary considering it's only comparing values supplied by the read loop.

I think we need to be able to advance `examinedSegment` and `examinedIndex` to subsequent `BufferSegment`s if the goal is to not throw for unexamining. If we want to go that route, I think it makes sense to store the entire `_lastExaminedSequencePosition` rather than just the `_lastExaminedIndex` and use that if it's greater than `consumed` which could also probably be done outside a lock.